### ISSUE TITLE
[TypeChecker] ICE in solc tangerineWhistle and spuriousDragon due to an attempted use of an invalid external function type in a call

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#10113]
## Description

```
contract C {
  function f(function() external returns (function())[] memory x) {
    x[]();
  }
}
```

throws

https://github.com/ethereum/solidity/blob/96c188be9dd790613a2b2c896e17b4cb1ae52a9e/libsolidity/ast/Types.cpp#L3005

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): tangerineWhistle, spuriousDragon (fixed in byzantium)

## Steps to Reproduce

```
$ solc --evm-version tangerineWhistle test.sol
```

**Comment by bshastry** [comment]
Please note that this can be triggered through free functions alone. Example

```
function f(function() external returns (function())[] memory x) {
  x[2]();
}
```

Edit: Invocation
```
$ solc --evm-version tangerineWhistle test.sol
```

**Comment by cameel** [comment]
It's also not related to arrays or function parameters at all. Here's a repro without these elements:

```solidity
contract C {
    function f() public {
        function() external returns (function() internal) getCallback;
        getCallback();
    }
}
```
```
Internal compiler error during compilation:
/solidity/libsolidity/ast/Types.cpp(2852): Throw in function solidity::frontend::TypePointers solidity::frontend::FunctionType::returnParameterTypesWithoutDynamicTypes() const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what:
[solidity::util::tag_comment*] =
```

The cause is simply that type checking does not stop when it determines that the return type is not valid. It's possible to reach a point where we need to actually use the type and `returnParameterTypesWithoutDynamicTypes()` fails because it has an assert against that.

**Comment by cameel** [comment]
I have a fix for this: #11478.
